### PR TITLE
Use for update skip locked for postgresql polling

### PR DESCRIPTION
**Pull request by gmokki**
Fixes a deadlock observed when multiple engines are running concurrently

### CODE REVIEW

File: nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java [hunk 1]
```diff
@@ -97,8 +97,8 @@ public boolean hasUpdateableCTE() {
     }
 
     @Override
-    public String forUpdateInnerSelect() {
-      return "";
+    public String forUpdateInnerSelect(boolean skipLocked) {
+        return skipLocked ? " with (updlock,readpast)" : " with (updlock)";
```
Comment (jsyrjala) [inline nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java:101]: This is always called with `true`. Is this conditional needed?

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,7 @@
   - Throw `IllegalArgumentException` instead of `IllegalStateException` when trying to update workflow instance state to an invalid value.
   - Throw `IllegalArugmentException` instead of `RuntimeException` when trying to insert workflow instance with unknown type or with a state that is not a start state.
   - Make `StateVariableTooLongException` extend `IllegalArgumentException` instead of `RuntimeException`.
+  - Fix sql deadlocks in polling for PostgreSQL with skip locked
```
Comment (efonsell) [inline CHANGELOG.md:15]: Seems like this is done for all databases? sql -> SQL

File: nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                .query("businessKey", UNIQUE_KEY)
+                .query("status", finished)
+                .query("include", "").get(ListWorkflowInstanceResponse[].class);
+        if (instances.length == WORKFLOWS) {
+          return instances;
+        }
+      }
+    });
+    assertThat(wfr.length, is(WORKFLOWS));
+    List<DateTime> startTimes = Stream.of(wfr).map(r -> r.started).sorted().collect(toList());
```
Comment (efonsell) [inline nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java:121]: startTimes is not used, can this be removed?

File: nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java (hunk 1 above)
Comment (gmokki) [reply nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java:101]: I think it is more clear this way. But I could rename the method to forUpdateInnerSelectSkipLocked and remove the boolean. Not sure if we ever need the false option

File: CHANGELOG.md (hunk 2 above)
Comment (gmokki) [reply CHANGELOG.md:15]: Other databases had a completely different fix already few releases ago. This code path is only for pg

File: nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java (hunk 3 above)
Comment (gmokki) [reply nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java:121]: I was thinking of coming up with something to test with the start times, comparing the max between avg. But could not get any difference with or without deadlocks. What would be a good way to test if the deadlock condition occured?

File: nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java (hunk 1 above)
Comment (efonsell) [reply nflow-engine/src/main/java/io/nflow/engine/config/db/SqlServerDatabaseConfiguration.java:101]: IMO remove boolean and rename if needed

File: CHANGELOG.md (hunk 2 above)
Comment (efonsell) [reply CHANGELOG.md:15]: Why there are changes in SqlServerDatabaseConfiguration then?

File: nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java (hunk 3 above)
Comment (efonsell) [reply nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java:121]: If there's no easy way to test it, just remove it?

File: CHANGELOG.md (hunk 2 above)
Comment (gmokki) [reply CHANGELOG.md:15]: The SQL server does support skip locked, and it uses it now in archiving code. But there should never have been any deadlocks there (or at least the archived rows should be so old as to not conflict with engine). The only database that uses the `select * from workflow where id in (update set executor where executor is null returning id)` is postgresql. it is the only sql variant that currently returns true for `hasUpdateReturning`. Technically at least ms sqlserver supports it too, but syntax is very different with `OUTPUT` so I have not taken it into use yet.

**Pull request merged**